segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/BaseChunkStorage.java

```diff
@@ -0,0 +1,563 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import io.pravega.common.LoggerHelpers;
+import io.pravega.common.Timer;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Base implementation of {@link ChunkStorage}.
+ * It implements common functionality that can be used by derived classes.
+ * Delegates to specific implementations by calling various abstract methods which must be overridden in derived classes.
+ * <div>
+ * Below are minimum requirements that any implementation must provide.
+ * Note that it is the responsibility of storage provider specific implementation to make sure following guarantees are provided even
+ * though underlying storage may not provide all primitives or guarantees.
+ * <ul>
+ * <li>Once an operation is executed and acknowledged as successful then the effects must be permanent and consistent (as opposed to eventually consistent)</li>
+ * <li>{@link ChunkStorage#create(String)}  and {@link ChunkStorage#delete(ChunkHandle)} are not idempotent.</li>
+ * <li>{@link ChunkStorage#exists(String)} and {@link ChunkStorage#getInfo(String)} must reflect effects of most recent operation performed.</li>
+ * </ul>
+ * </div>
+ * <div>
+ * There are a few different capabilities that ChunkStorage may provide.
+ * <ul>
+ * <li> Does {@link ChunkStorage} support appending to existing chunks?
+ * This is indicated by {@link ChunkStorage#supportsAppend()}. For example S3 compatible Chunk Storage this would return false. </li>
+ * <li> Does {@link ChunkStorage}  support for concatenating chunks? This is indicated by {@link ChunkStorage#supportsConcat()}.
+ * If this is true then concat operation concat will be invoked otherwise append functionality is invoked.</li>
+ * <li>In addition {@link ChunkStorage} may provide ability to truncate chunks at given offsets (either at front end or at tail end). This is indicated by {@link ChunkStorage#supportsTruncation()}. </li>
+ * </ul>
+ * There are some obvious constraints - If ChunkStorage supports concat but not natively then it must support append .
+ *
+ * For concats, {@link ChunkStorage} supports both native and append, ChunkManager will invoke appropriate method depending on size of target and source chunks. (Eg. ECS)
+ * </div>
+ *
+ * <div>
+ * The implementations in this repository are tested using following test suites.
```

Comment:
This is given as a guidance that  derived classes should be tested using this set of tests. This is helpful reminder for anyone implementing this class.
We don't have a separate compatibility test suite as of now.

Reply:
I think that any developer is able to locate where a given class is being tested, right? I'm not saying that this information is not useful, what I'm saying is that, as far as I'm aware, nowhere else in the Javadoc we enumerate the test classes related to the class, so this is a comment about consistency. Also, I suspect that this may introduce additional maintenance burden, as we need to keep this part of the Javadoc updated for any changes related to test classes.

Reply:
This is intentional.